### PR TITLE
Add local registry overlays

### PR DESCRIPTION
**Pull request by jneem**
This PR adds (private to cargo internals) support for local registry overlays, in which you can locally pretend to add packages to remote registries; the local packages will have the same source ids as the remote registry that you're overlaying.

There are two ways to set up these overlays: programmatically using `GlobalContext::local_overlays` and through the `__CARGO_TEST_PACKAGE_CONFUSION_VULNERABILITY_DO_NOT_USE_THIS` environment variable. You can't set up these overlays with `.cargo/config`.

The motivation for this is [packaging workspaces](https://github.com/rust-lang/cargo/issues/10948). When we're packing a workspace, we'd like to be able to pretend (for lockfile generation and verification) that some workspace packages are already published even though they aren't.

### CODE REVIEW

File: crates/cargo-test-support/src/lib.rs [hunk 1]
```diff
@@ -844,6 +844,17 @@ impl Execs {
         self
     }
 
+    pub fn overlay_registry(&mut self, url: &Url, path: &str) -> &mut Self {
+        if let Some(ref mut p) = self.process_builder {
+            let env_value = format!("{}={}", url, path);
+            p.env(
+                "__CARGO_TEST_PACKAGE_CONFUSION_VULNERABILITY_DO_NOT_USE_THIS",
```
Comment (Eh2406) [inline crates/cargo-test-support/src/lib.rs:851]: Nit: The term of art for this kind of vulnerability is "Dependency Confusion". As with so many names in computer science, it's better to have consistency then trying to find a better name. This term was coined in the blog post https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610 which might be worth linking in comments about how users could easily miss use this.

File: src/cargo/sources/overlay.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    }
+}
+
+impl<'gctx> Source for OverlayConfusionAttack<'gctx> {
+    fn source_id(&self) -> crate::core::SourceId {
+        self.remote.source_id()
+    }
+
+    fn supports_checksums(&self) -> bool {
+        true
```
Comment (Eh2406) [inline src/cargo/sources/overlay.rs:32]: I would've guessed that this depends on whether both of the sources being overlaid support checksum. Did I miss something?

File: src/cargo/sources/overlay.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    fn requires_precise(&self) -> bool {
+        false
```
Comment (Eh2406) [inline src/cargo/sources/overlay.rs:36]: I would've thought this was true if either of the sources required precise, what am I missing?

File: src/cargo/sources/overlay.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        let local_source = self.local.source_id();
+        let remote_source = self.remote.source_id();
+
+        let local_dep = dep.clone().map_source(remote_source, local_source);
+        let mut callback = |index: IndexSummary| {
+            let index = index.map_summary(|s| s.map_source(local_source, remote_source));
+            f(index)
+        };
+        ready!(self.local.query(&local_dep, kind, &mut callback))?;
+        self.remote.query(dep, kind, f)
```
Comment (Eh2406) [inline src/cargo/sources/overlay.rs:54]: Does this need to handle the same package/version being available from both registries? Possibly given its limited use case, just by detecting the situation and crashing.

**Comment by Eh2406** [comment]
Catching Eric up on some of the context, Arlo and I discussed the questions from https://github.com/rust-lang/cargo/issues/10948#issuecomment-2112683333 at office hours on Thursday. Adding this API, exclusively for internal use, seemed like the least hacky solution.

File: src/cargo/sources/overlay.rs (hunk 4 above)
Comment (jneem) [reply src/cargo/sources/overlay.rs]: Good point, I've added that check and a test.

File: src/cargo/sources/overlay.rs (hunk 2 above)
Comment (jneem) [reply src/cargo/sources/overlay.rs]: Good point, I'd had in my head (and in the documentation) that the two sources were both `RegistrySource`s. I've updated the docs and this method to reflect that they could be arbitrary `Source`s.

File: src/cargo/sources/overlay.rs (hunk 3 above)
Comment (jneem) [reply src/cargo/sources/overlay.rs]: (As above)

File: src/cargo/sources/overlay.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
+use std::task::ready;
+
+use crate::sources::IndexSummary;
+
+use super::source::{MaybePackage, Source};
+
+/// A `Source` that overlays a one source over another, pretending that the packages
```
Comment (torhovland) [inline src/cargo/sources/overlay.rs:7]:
```suggestion
/// A `Source` that overlays one source over another, pretending that the packages
```

File: src/cargo/util/context/mod.rs [hunk 6]
```diff
@@ -523,6 +526,13 @@ impl GlobalContext {
             .borrow_mut()
     }
 
+    /// Local overlay configuration.
+    ///
+    /// See [`crate::sources::overlay::OverlayConfusionAttack`] for why you shouldn't use this.
+    pub fn local_overlays(&self) -> RefMut<'_, HashMap<SourceId, PathBuf>> {
```
Comment (torhovland) [inline src/cargo/util/context/mod.rs:532]: Should probably mention that the paths need to point to local registries, not just to any crate file or directory.

File: src/cargo/sources/overlay.rs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+/// available in the overlay are actually available in the other one.
+///
+/// This is a massive footgun and a terrible idea, so we do not (and never will)
+/// expose this publicly. However, it is useful for some very specific private
+/// things, like locally verifying a bunch of packages at a time before any of
+/// them have been published.
+pub struct OverlayConfusionAttack<'gctx> {
```
Comment (torhovland) [inline src/cargo/sources/overlay.rs:14]: Frankly, I think `OverlaySource` is a better fitting name 🙂

File: Cargo.toml [hunk 8]
```diff
@@ -60,7 +60,7 @@ indexmap = "2.2.6"
 itertools = "0.12.1"
 jobserver = "0.1.31"
 lazycell = "1.3.0"
-libc = "0.2.154"
+libc = "0.2.153"
```
[the remark this answers is not in the document]
Comment (jneem) [reply Cargo.toml:63]: 0.2.154 got [yanked](https://github.com/rust-lang/libc/issues/3677#issuecomment-2093991553). I see that 0.2.155 just got released, so I'll switch to that

File: src/cargo/sources/overlay.rs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+impl<'gctx> OverlayConfusionAttack<'gctx> {
+    pub fn new(local: Box<dyn Source + 'gctx>, remote: Box<dyn Source + 'gctx>) -> Self {
+        Self { local, remote }
+    }
+}
+
+impl<'gctx> Source for OverlayConfusionAttack<'gctx> {
+    fn source_id(&self) -> crate::core::SourceId {
+        self.remote.source_id()
```
[the remark this answers is not in the document]
Comment (jneem) [reply src/cargo/sources/overlay.rs:30]: In the current implementation that can't happen because the overlays are configured by a `HashMap<SourceId, _>`, so only one overlay can be registered for each remote.

File: src/cargo/sources/overlay.rs (hunk 7 above)
Comment (jneem) [reply src/cargo/sources/overlay.rs:14]: That name definitely makes more sense, yes. But in office hours someone suggested giving it a scary-looking name to make people less likely to misuse it...

File: tests/testsuite/registry_overlay.rs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    // The one we want is in the main registry.
+    Package::new("baz", "0.1.1").publish();
+    // An older version in the local registry doesn't confuse it.
+    Package::new("baz", "0.0.1")
+        .alternative(true)
+        .local(true)
+        .publish();
```
Comment (epage) [inline tests/testsuite/registry_overlay.rs:68]: I don't think this is really testing anything because `baz` is too low of a version for the version requirement

File: tests/testsuite/registry_overlay.rs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        .overlay_registry(&reg.index_url(), &alt_path)
+        .run();
+}
+
+#[cargo_test]
+fn version_fallback() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
```
Comment (epage) [inline tests/testsuite/registry_overlay.rs:61]: What is the role of `bar` in these tests?   We should either simplify the test cases or name the packages to convey the role they play in the tests

File: tests/testsuite/registry_overlay.rs [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+fn multiversion() {
+    let (reg, alt_path) = setup();
+    let p = simple_project();
+
+    Package::new("bar", "0.0.1").publish();
+    Package::new("baz", "0.1.1").dep("baz", "=0.0.1").publish();
+    Package::new("baz", "0.0.1")
+        .alternative(true)
+        .local(true)
+        .publish();
```
Comment (epage) [inline tests/testsuite/registry_overlay.rs:115]: Should we have the local / selected version depend on the registry version?  Thats more like what we'll be dealing with for `cargo package`

File: tests/testsuite/registry_overlay.rs [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                edition = "2015"
+                authors = []
+
+                [dependencies]
+                bar = ">= 0.0.0"
+                baz = ">= 0.1.0"
```
Comment (epage) [inline tests/testsuite/registry_overlay.rs:32]: Why are non-standard version reqs being used?  Chekhov's gun would say that there is something specific about this to the tests but I'm not seeing it.

(ditto for the other package)

File: src/cargo/util/context/mod.rs [hunk 14]
```diff
@@ -523,6 +526,16 @@ impl GlobalContext {
             .borrow_mut()
     }
 
+    /// Local overlay configuration.
+    ///
+    /// The values in this map are paths pointing to a local registry. Inserted overlays will
+    /// affect any newly created `PackageRegistry`s.
+    ///
+    /// See [`crate::sources::overlay::OverlayConfusionAttack`] for why you shouldn't use this.
+    pub fn local_overlays(&self) -> RefMut<'_, HashMap<SourceId, PathBuf>> {
```
Comment (epage) [inline src/cargo/util/context/mod.rs:535]: I'd hope we can avoid putting this on GlobalContext but its hard to say without reviewing the follow up at the same time.  Should we defer this to the follow up PR?

My motivation is that it gets messy adding and removing from the Global Context and if we can directly instantiate things, the easier it will be to maintain.

File: src/cargo/sources/overlay.rs (hunk 7 above)
Comment (epage) [reply src/cargo/sources/overlay.rs:16]: I like `OverlaySource` being the suffix.  Maybe `DependencyConfusionThreatOverlaySource`?

File: src/cargo/sources/config.rs [hunk 15]
```diff
@@ -136,7 +170,8 @@ impl<'gctx> SourceConfigMap<'gctx> {
         debug!("loading: {}", id);
 
         let Some(mut name) = self.id2name.get(&id) else {
-            return id.load(self.gctx, yanked_whitelist);
+            let src = id.load(self.gctx, yanked_whitelist)?;
+            return self.overlay_source(id, src, yanked_whitelist);
```
Comment (epage) [inline src/cargo/sources/config.rs:174]: We only call `overlay_source` on the result of a `load`

Should we name it `fn load` and perform the load and apply the overlay?  I worry that the current structure could make it easy to overlook applying the overlay while this pattern would make it the default to go through `Self::load` and we'd be less likely to introduce bugs to this

File: src/cargo/sources/config.rs [hunk 16]
```diff
@@ -81,6 +84,22 @@ impl<'gctx> SourceConfigMap<'gctx> {
...
+
+        if let Ok(overlay) =
+            gctx.get_env("__CARGO_TEST_DEPENDENCY_CONFUSION_VULNERABILITY_DO_NOT_USE_THIS")
+        {
+            let (url, path) = overlay.split_once('=').ok_or(anyhow::anyhow!(
+                "invalid overlay format. I won't tell you why; you shouldn't be using it anyway"
+            ))?;
+            let overlay_id = SourceId::for_local_registry(path.as_ref())?;
+            base.overlays.insert(SourceId::from_url(url)?, overlay_id);
+        }
```
Comment (epage) [inline src/cargo/sources/config.rs:101]: Should we move this hack up into the caller?  `PackageRegistry` is the immediate caller but maybe a layer above that?  That would help with the `GlobalContext`.  The main downside is it would make it more annoying to support `__CARGO_TEST_DEPENDENCY_CONFUSION_VULNERABILITY_DO_NOT_USE_THIS`

Thoughts?

File: src/cargo/util/context/mod.rs (hunk 14 above)
Comment (jneem) [reply src/cargo/util/context/mod.rs:535]: I can give a little more context before we decide whether or not to defer. `cargo package` needs the overlay to take effect during a call to `ops::compile`. AFAICT there are three possibilities:

- put the overlay configuration in `GlobalContext`
- put the overlay configuration in the ephemeral workspace that `cargo package` creates
- have `cargo package` do something lower-level than `ops::compile`

The workspace option has the advantage that the overlay configuration doesn't need to be removed (because the workspace is discarded). It seems like it would just require a little work in `resolve_ws` and `resolve_ws_with_opts` to get the overlay configuration out of the workspace and insert it into the `PackageRegistry`

File: tests/testsuite/registry_overlay.rs (hunk 13 above)
Comment (jneem) [reply tests/testsuite/registry_overlay.rs:32]: I think I just copied it from some other test. I'll switch to normal version reqs.

File: tests/testsuite/registry_overlay.rs (hunk 11 above)
Comment (jneem) [reply tests/testsuite/registry_overlay.rs:61]: I think it's not needed; I'll remove it.

File: tests/testsuite/registry_overlay.rs (hunk 10 above)
Comment (jneem) [reply tests/testsuite/registry_overlay.rs:68]: The motivation here was to test that the overlay is correctly the union of the versions in the overlay and the overlaid registries. (If it were only returning the overlay versions, it would only return the too-old version and fail to resolve)

File: tests/testsuite/registry_overlay.rs (hunk 12 above)
Comment (jneem) [reply tests/testsuite/registry_overlay.rs:115]: I don't think so, but also happy to discuss it further. My understanding was that the behavior of `cargo package` w.r.t. intra-workspace deps should be exactly the same as if the deps were published before packaging. Resolution should have the option to select the versions that are being packaged simultaneously, but it's also allowed not to.

File: src/cargo/sources/config.rs (hunk 15 above)
Comment (jneem) [reply src/cargo/sources/config.rs:174]: Unfortunately we're already in a `fn load`. I renamed `overlay_source` to `load_and_overlay`, and modified it to do the load and the overlay. I think this is an improvement but not sure it it really addresses the concern.

File: src/cargo/sources/config.rs (hunk 16 above)
Comment (jneem) [reply src/cargo/sources/config.rs:101]: I think if we moved the overlay configuration into `Workspace` as discussed below then this could be pushed up into `resolve_ws` and `resolve_ws_with_opts`

File: src/cargo/util/context/mod.rs (hunk 14 above)
Comment (epage) [reply src/cargo/util/context/mod.rs:535]: The workspace seems the most beneficial.  Whichever way we go, still seems like it could be split out into a follow up PR

File: tests/testsuite/registry_overlay.rs (hunk 12 above)
Comment (epage) [reply tests/testsuite/registry_overlay.rs:110]: Yes, that should be the behavior but we aren't testing that situation to confirm it.  `cargo package` will involve a local package depending on an overlayed package which may depend on a registry package.

I guess there are also situations where you depend on a registry package which will depend on the overlaid package (if `clap_complete` used `criterion` it would both depend on clap directly and through `criterion`).  So maybe its best to cover both cases.

File: tests/testsuite/registry_overlay.rs [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+        alt.index_url()
+            .to_file_path()
+            .unwrap()
+            .into_os_string()
+            .into_string()
+            .unwrap(),
+    )
+}
+
+fn simple_project() -> Project {
```
Comment (epage) [inline tests/testsuite/registry_overlay.rs:19]: Let's avoid using this.  This reduces "duplication" but that is false code duplication.  They have the appearance of being the same but they do not share requirements.  Instead it affects the code locality (the ability to read tightly coupled code).  I'm trying to double check what these tests are doing and jumping back and forth is making this more difficult.

File: tests/testsuite/registry_overlay.rs (hunk 10 above)
Comment (epage) [reply tests/testsuite/registry_overlay.rs:65]: Except we aren't checking the resolve error for whether it sees `baz@0.0.1`

We also should probably have a test the other direction, that the overlay source properly shadows the regular source.

File: tests/testsuite/registry_overlay.rs (hunk 12 above)
Comment (jneem) [reply tests/testsuite/registry_overlay.rs:110]: Ah, makes sense, I just misunderstood your original comment. I'm now testing both directions.

> I guess there are also situations where you depend on a registry package which will depend on the overlaid package
I think the registry package will never *require* an overlaid package, right? Because the registry package can satisfy all its dependencies from the registry. I added a test where the registry package will pick up an overlaid dependency because it's newer (but compatible with) the registry version of that dependency.

File: tests/testsuite/registry_overlay.rs (hunk 10 above)
Comment (jneem) [reply tests/testsuite/registry_overlay.rs:65]: Got it. I'm now testing both directions, and checking the appropriate versions in the stderr.

File: Cargo.toml [hunk 18]
```diff
@@ -60,7 +60,7 @@ indexmap = "2.2.6"
 itertools = "0.12.1"
 jobserver = "0.1.31"
 lazycell = "1.3.0"
-libc = "0.2.154"
+libc = "0.2.155"
```
Comment (epage) [inline Cargo.toml:63]: Why is libc being bumped as part of this PR?

File: tests/testsuite/registry_overlay.rs (hunk 12 above)
Comment (epage) [reply tests/testsuite/registry_overlay.rs]: Not require but should resolve to use it.

File: src/cargo/sources/config.rs [hunk 19]
```diff
@@ -125,6 +145,22 @@ impl<'gctx> SourceConfigMap<'gctx> {
         self.gctx
     }
 
+    fn load_and_overlay(
```
Comment (epage) [inline src/cargo/sources/config.rs:148]: nit: I feel like `and` in a name is a code smell, that you are making a fucntion do unrelated things.  Ideally, we find a name that gets at the heart of what distinguishes this from `load`

Comment (epage) [inline src/cargo/sources/config.rs:148]: nit: I tend to prefer putting internal functions after their use.  People tend to read top-down and tend to be looking at interfaces and then look into implementations where relevant.

File: Cargo.toml (hunk 18 above)
Comment (jneem) [reply Cargo.toml]: I've removed it. (It had something to do with 0.2.154 being yanked, but I don't remember anymore why that was a problem)

File: src/cargo/sources/config.rs (hunk 19 above)
Comment (jneem) [reply src/cargo/sources/config.rs]: What about `load_unreplaced`? Because the difference between `load` and `load_unreplaced` is that `load` does source replacement.

Comment (jneem) [reply src/cargo/sources/config.rs]: Done

Comment (epage) [reply src/cargo/sources/config.rs]: Describing something by what it isn't doing is a bit messy.

Your name hints that `pub fn load` is `load_replaced`.  So maybe this could be `load_overlaid`.  Still not thrilled but probably the best we can do without blocking this further.

File: src/cargo/sources/config.rs (hunk 16 above)
Comment (epage) [reply src/cargo/sources/config.rs:101]: Just wanted to check in if there is anything you are waiting on me for this change as you've done other work on this PR but not this part yet.

**Comment by epage** [comment]
btw feel free to clean up your commit history for how you want to present it for review and merge.

File: src/cargo/sources/config.rs (hunk 16 above)
Comment (jneem) [reply src/cargo/sources/config.rs:101]: Ah sorry, I thought this would go in a follow-up (along with moving the overlay configuration to `Workspace`). On Monday I'll clean up the commit history and look into doing this too.

File: src/cargo/ops/resolve.rs [hunk 20]
```diff
@@ -116,7 +116,8 @@ version. This may also occur with an optional dependency that is not enabled.";
 /// This is a simple interface used by commands like `clean`, `fetch`, and
 /// `package`, which don't specify any options or features.
 pub fn resolve_ws<'a>(ws: &Workspace<'a>, dry_run: bool) -> CargoResult<(PackageSet<'a>, Resolve)> {
-    let mut registry = PackageRegistry::new(ws.gctx())?;
+    let source_config = SourceConfigMap::new_with_overlays(ws.gctx(), ws.local_overlays()?)?;
+    let mut registry = PackageRegistry::new_with_source_config(ws.gctx(), source_config)?;
```
Comment (epage) [inline src/cargo/ops/resolve.rs:120]: How do we decide when to replace `PackageRegistry::new` with this?

We only replaced two uses of it:
```
src/cargo/ops/resolve.rs
119:    let mut registry = PackageRegistry::new(ws.gctx())?;
145:    let mut registry = PackageRegistry::new(ws.gctx())?;

src/cargo/ops/cargo_package.rs
475:    let mut tmp_reg = PackageRegistry::new(ws.gctx())?;

src/cargo/ops/cargo_update.rs
35:    let mut registry = PackageRegistry::new(ws.gctx())?;
76:                    let mut registry = PackageRegistry::new(opts.gctx)?;
91:    let mut registry = PackageRegistry::new(opts.gctx)?;
229:    let mut registry = PackageRegistry::new(gctx)?;

src/cargo/ops/cargo_add/mod.rs
81:    let mut registry = PackageRegistry::new(options.gctx)?;

crates/xtask-bump-check/src/xtask.rs
393:    let mut registry = PackageRegistry::new(gctx)?;
```

Comment (epage) [inline src/cargo/ops/resolve.rs:120]: Depending on which type is more correct to know about the other, should we have either
- `Workspace::package_registry(&self) -> PackageRegistry`
- `PackageRegistry::with_workspace(ws: &Workspace) -> Self`

to simplify this pattern and make it more discoverable so that when new features get added, people are more likely to use this and not break support for this in `cargo package` in odd corner cases?

File: crates/cargo-test-support/src/registry.rs [hunk 21]
```diff
@@ -1649,7 +1649,12 @@ impl Package {
     /// Returns the path to the compressed package file.
     pub fn archive_dst(&self) -> PathBuf {
         if self.local {
-            registry_path().join(format!("{}-{}.crate", self.name, self.vers))
+            let path = if self.alternative {
```
Comment (epage) [inline crates/cargo-test-support/src/registry.rs:1652]: Please move this to the appropriate commit

File: src/cargo/ops/resolve.rs (hunk 20 above)
Comment (jneem) [reply src/cargo/ops/resolve.rs]: I don't have a great sense for which is better, but since `Workspace` already mentions `PackageRegistry` (and `PackageRegistry` doesn't mention `Workspace`) I went with `Workspace::package_registry`. All the usages of `PackageRegistry::new` had a workspace available, so I converted them all and then deleted `PackageRegistry::new` so that future changes are likely to default to `Workspace::package_registry`

**Comment by epage** [comment]
Thanks!

@bors r+
